### ISSUE TITLE
Feature related issuue

### ISSUE THREAD

**Issue by poseidon1214** [JohnLangford/vowpal_wabbit#857]
1. When Using hash(-b parameter), we should know the feature count in the training data. However, my training dataset will increase daily, a fixed [-b] parameter will cause the feature collision.
2. I set the (-b parameter) a little large to avoid feature collision, but the L1 feature selection doesn't make most features to be around zero, why?
 Any idea could help me?  Great thanks~

**Comment by arielf** [comment]
`--l1` isn't supposed to make features be around zero.

`--l1` should actually do the opposite: trim features whose weight is near zero (lower than the `--l1` value) and leave the more significant values in the model.

**Comment by poseidon1214** [comment]
yes，but my features' weight most are not around zero when using --l1. It means that most of my features are useful? The feature values in the training data are very sparse, I guess there must be some useless features in the training data. However, the --l1 experiment result shows that almost all features are useful(--l1 not very small). Would you have some suggestions?

**Comment by arielf** [comment]
How wide is the data set?
The more features you have (per example) the lower their weights (vs the constant) would be.
`--l1 0.001` looks still like a very large value.  How many features are left in the model after applying it?

**Comment by arielf** [comment]
Not sure I understand what the problem is, and why this is filed as an issue with `vw`.

`vw` is streaming/tracking, meaning that even if you have new features over time, and even if there are collisions with older features, the old features, that no longer appear, would be gradually forgotten over time and the weights in collision slots will gravitate towards the current prevalent features weights.  You could use `--power_t` smaller than the default 0.5 to "forget the past" faster (be careful to not over do) but that may actually hurt if very old features which seldom repeat are in fact important and you don't want to forget them fast.

Without seeing your data profile, and since it is not clear what exactly you mean by "new features over time," are they replacing older feature or get added? If they get added, does this mean the width of the data set (# of features per example) grows with time? It is hard for me to say.  What's the number of features per example?  What's the number of distinct features per N examples?  Is this a real, or just a perceived problem?  Are you actually seeing more features in the readable model than your preset `-b <bits>` space?  If you do, you should increase `--l1` until you don't.

"Almost all features are useful" is typical when the data is very narrow and features almost never repeat.  You can somewhat mitigate for reaching high weights quickly, by learning slower (using a lower learning rate e.g `-l 0.1`).

**Comment by poseidon1214** [comment]
I have  20 million instances and the number of features is about 60 million.
When L1 =1.0, all the feature weights are nonzero. When L1=1e-10, the dimension of features whose weights are nonzero is 1.4 million. 
I guess that VW trim features whose weight is lower than the --l1 value, at the same time, the weight is normalized. Is that right ?
Greaaaaaaaat thanks.

**Comment by JohnLangford** [comment]
l1 regularization applies to the gradient on a per-example basis.  
Hence, values like 1e-4 are common.  I'm unclear on why all features 
have nonzero weights when l1=1.0.  That sounds aberrant, and an example 
would be handy.

-John

**Comment by poseidon1214** [comment]
thanks~
